### ISSUE TITLE
How to load into browser environment via script tag?

### ISSUE THREAD

**Issue by pqr** [ai/nanoid#127]
Not clear how to run **nanoid** within browser, which file should I load via script tag?
Are there `dist` or `.min.js` build available?

Think this information would be valuable for documentation (README.md) as well, because right now all examples show only Node.js `require` synax.

**Comment by ai** [comment]
Right now we do not provide this method by design. Using CDN with `.min.js` version is very bad for user performance.

I recommend to tale Parcel and use Nano ID from npm. Parcel is very easy to use tool with great DX.

But if you have a good use case for `.min.js` I can reconsider (I will re-open issue).

**Comment by franciscop-sc** [comment]
👍  This would also be useful for generating N IDs randomly for one-off cases without having to install it locally and compile it into a browser bundle. Or for small demos where it is loading 1-3 script tags manually.

**Comment by ai** [comment]
@franciscop-sc can you use https://unpkg.com/ for this?

**Comment by ai** [comment]
@franciscop-sc a few months later I will reinvestigate ES modules status (right now there are a lot of blank zones there) and will try to create something which can work in browsers.

Right now you can use Parcel. It has a great DX even for small hacks.

**Comment by tatomyr** [comment]
I'm not sure regarding loading via `script` tag, but having the possibility to use native browser imports for **nanoid** would be a really great feature. Please consider exposing a bundled file somewhere.

**Comment by ai** [comment]
@tatomyr unfortunately native imports are not ready. We need to fix a lot of issue before using them. Without solutions of these issues, we don’t know how to use imports properly.

My favorite example that all open source with `import lodash from 'lodash'` is not native imports, since native imports need a direct path to the file (`from '../../node_modules/lodash/lodash.js'`).

This is why I am against hype drive development and prefer to wait good ESM based bundler before using ESM.

**Comment by franciscop-sc** [comment]
Agreed, I might have jumped a bit early into ESM in my personal projects (this is a work account). But someone has to jump in first and find all the bugs, right? Which I'm definitely am.

BTW, thanks for nanoid, I love it.

**Comment by tatomyr** [comment]
@ai actually **npm** is what I'm trying to avoid. What I'd like to write, is either bare `import nanoid from 'https://.../nanoid.js` or using proxy `deps` file in the root of my project.

**Comment by ai** [comment]
@tatomyr what bundler do you use to about round-trip latency chain?

For instance, you have `app.js`, it imports `liba.js`, it imports `libb.js`. Modern phone will not be able to download it faster than 1,5s because every request requires 200-500ms just for round-trip. It is huge problem for performance.

If you use different domain for your libraries it makes everything even worse, since it will be out of HTTP/2 channel of your domain (note, that HTTP/2 doesn't solve round-trip latency problem out of the box).

**Comment by tatomyr** [comment]
@ai thanks for your explanation. 

However, at the moment it's not an issue for me since I'm building PWA and cashing every dependency. Luckily **nanoid** doesn't depend on other packages.

Also, I'd like to use **nanoid** for my **deno** projects, hence I don't need a bundler.

Hope I've managed to clarify my point.

**Comment by ai** [comment]
I do not want to put build files to npm, since it will increase npm package size.

Do you know any CDN where we can put custom files?

**Comment by chigix** [comment]
@ai increase of npm package size should not and would not be encountered as a problem. Because all npm package are managed under a `node_modules` directory in a project, no matter node program or front-end project such as Angular, it is the `node_modules` directory that not designed for an online web path hierarchy, which makes all front-end project using npm need to build up another online directory organization even webpack is not employed.

For node problem, it seems that there is originally no size problem for `nanoid`.

**Comment by chigix** [comment]
[unpkg](https://unpkg.com/) might be a good option for releasing min.js distribution file through cdn.

**Comment by TrySound** [comment]
Unpkg with `module` flag converts esm packages into embeddable modules.
https://unpkg.com/jss@10.0.0-alpha.17/dist/jss.esm.js?module

**Comment by franciscop-sc** [comment]
You can use `package.json`'s "files" key to include only wanted files, or `.npmignore` to exclude unwaanted files. But you *probably* want to include the minified file in the package, since it's the only way to track versions...

There are few services that work straight with npm:
- https://unpkg.com/nanoid@2.0.3/index.js
- https://cdn.jsdelivr.net/npm/nanoid@2.0.3

**Comment by franciscop-sc** [comment]
Another alternative I've seen is to setup a `-dist` package like `nanoid-dist` and then publish the minified files there. But then you have to track/maintain both packages.

**Comment by chigix** [comment]
Elasticsearch releasing performs like the pattern @franciscop-sc  mentioned:
* https://www.npmjs.com/package/elasticsearch-browser
* https://yarnpkg.com/en/package/@elastic/elasticsearch

**Comment by ai** [comment]
> increase of npm package size should not and would not be encountered as a problem

Just check the actual size of `node_modules` any project. It is a huge.

And we have so big `node_modules` because we put a lot of unnecessary things there.

It is already a problem. Just not a hype problem to read it in Twitter.

**Comment by ai** [comment]
> You can use package.json's "files" key to include only wanted files, or .npmignore to exclude unwanted files.

We already did it of course. We went even further. We use `npx clean-publish` to release Nano ID and cut development configs from `package.json`.

**Comment by makoven** [comment]
> We can use Rollup and create some small files just for the demo.

Nope you can't. Rollup creates bundle, but browser throws error at runtime: "random is not a function"

After adding "browser: true" in rollup config browser throws other error: "process is not defined"

**Comment by TrySound** [comment]
@makoven Use https://github.com/rollup/rollup-plugin-replace to achieve the same as with webpack. This is widely used convention. No reason to avoid it. To import modules in browsers directly you can always define own process global variable.
```
globalThis.process = {
  env: {
    NODE_ENV: 'development'
  }
}
```

**Comment by makoven** [comment]
@TrySound, es modules is also a widely used convention. Much more common than `process.env`
Sorry, but i don’t want to drag two plugins (commonjs and replace) into my project because of the tiny id generator

**Comment by feikeq** [comment]
In other words, you cannot use the script tag to load the Nano ID in a web page? I still can only use JS time for UUID?   @ai

**Comment by ai** [comment]
@feikeq you need to use bundler like Rollup or webpack. Having a separated `<script>` for each library is bad for performance.

We are now working on ESM version, but it will be useful only for small experiments.

**Comment by lili21** [comment]
why the index.browser.js file contains `if (process.env.NODE_ENV !== 'production')` ?

**Comment by makoven** [comment]
@lili21 Psst! Don't you want to try your own handmade id generator?

```typescript
const chars =
  '1234567890' +
  'abcdefghijklmnopqrstuvwxyz' +
  'ABCDEFGHIJKLMNOPQRSTUVWXYZ';

const charsLen = chars.length; // 62

export function genId(length = 21): string {
  const id: string[] = []
  for (let i = 0; i < length; i++) {
    id.push(chars[(Math.round(Math.random() * 100)) % charsLen]);
  }
  return id.join('');
}

```

**Comment by TrySound** [comment]
Here's the same. As you can see it's placed in non-secure folder.
https://github.com/ai/nanoid/blob/master/non-secure/index.js

**Comment by ai** [comment]
I created a tool for myself to publish Nano ID as ESM
https://github.com/ai/dual-publish/

I will test it on NanoEvents and will start preparing next major release for Nano ID.

**Comment by makoven** [comment]
@ai If you want, you can also consider supporting Deno users (just put a ".js" at the end of the each import statement)

**Comment by ai** [comment]
@makoven dual-publish automatically adds `index.mjs` path to improrts

**Comment by makoven** [comment]
@ai it seems node.js are going to abandon the *.mjs in future releases

**Comment by TrySound** [comment]
@makoven Where do you see this information? Could you provide a link?

**Comment by makoven** [comment]
@TrySound 
https://youtu.be/j86fFvEi9YQ?t=466
Not the most reliable source, I admit :) I asked him for a source. I will write if he answers

**Comment by TrySound** [comment]
> почти отказались

Это значит пересмотрели: добавили поле `type: commonjs | module` в package.json и .cjs расширение. Никто отказываться не собирается.

**Comment by ai** [comment]
I released [Nano Events](https://github.com/ai/nanoevents/) as an experiment of my new dual ESM/CJS system.

Does everyone like how it work? Nano ID will use the same system.

**Comment by makoven** [comment]
@ai Cool. It works. Wouldn’t it be wiser, for the happiness of future generations of developers, to declare the superiority of ESM? Claim **index.js** as an ESM (with {type: module} in **package.json**). And let file **index.cjs** hang around

**Comment by ai** [comment]
> Wouldn’t it be wiser, for the happiness of future generations of developers, to declare the superiority of ESM?

Hm, I try to make some experiments.

**Comment by ai** [comment]
@makoven I released Nano Evets 4.0. What do you think?

https://github.com/ai/nanoevents/#es-modules

**Comment by makoven** [comment]
@ai Checked for combinations with Typescript + Rollup. It bundles successfully in IIfe blob without a commonjs plugin! It would be awesome to have the same for storeon. 
But I can give a little criticism if you want. The default imports are evil. IMHO it would be better for something like that:
```javascript
import createNanoEvents from 'nanoevents';
const { createNanoEvents } = require('nanoevents');
```

**Comment by ai** [comment]
@makoven `import { createNanoEvents } from 'nanoevents';`

I like it too, but it requires Condition Exports. Unfortunately, Node.js will print warning. More details in my thread about ESM: https://twitter.com/sitnikcode/status/1229806223098306564

**Comment by makoven** [comment]
@ai I agree. But for now we have a separate file for the node, right? And ESM file, which is for browsers, bundlers, deno and typescript.

**Comment by ai** [comment]
@makoven we have a lot of Node.js users. For instance, I use Nano ID in my Logux Server and Logux’s users will see this warning too,

We can’t afford a warning for them on any `import` or `require()`.

**Comment by ai** [comment]
The ESM support was released in Nano ID 3.0

```js
import { nanoid } from 'https://cdn.jsdelivr.net/npm/nanoid/nanoid.js'
```
